File: Leetcode/Linked List/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse_list(self, head):
        if head is None:
            return head

        previous_node = None
        current_node = head

        while current_node:
            next_node = current_node.next
            current_node.next = previous_node
            previous_node = current_node
            current_node = next_node

        return current_node

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Grokking/Inplace Reversal/reverseLinkedList.js
class Node {
  constructor(value, next=null){
    this.value = value;
    this.next = next;
  }

  get_list() {
    result = "";
    temp = this;
    while (temp !== null) {
      result += temp.value + " ";
      temp = temp.next;
    }
    return result;
  }
};


const reverse = (head) => {
  // TODO: Write your code here
  let prev = null;
  let curr = head;

  while (curr) {
    const temp = curr.next;
    curr.next = prev;
    prev = curr;
    curr = temp;
  }
  return prev;
};

// Time Complexity: O(n), where n is the number of nodes within the linked list
// Space Complexity: O(1)